File: services/notification-service/src/Services/NotificationService.cs
using OtterWorks.NotificationService.Models;

namespace OtterWorks.NotificationService.Services;

public class NotificationService : INotificationService
{
    private readonly INotificationRepository _repository;
    private readonly ILogger<NotificationService> _logger;

    public NotificationService(
        INotificationRepository repository,
        ILogger<NotificationService> logger)
    {
        _repository = repository;
        _logger = logger;
    }

    public async Task ProcessEventAsync(SqsNotificationMessage sqsEvent)
    {
        var targetUserId = ResolveTargetUserId(sqsEvent);
        if (string.IsNullOrWhiteSpace(targetUserId))
        {
            _logger.LogWarning("No target user for event: {EventType}", sqsEvent.EventType);
            return;
        }

        var preferences = await _repository.GetPreferencesAsync(targetUserId);
        List<DeliveryChannel> enabledChannels;
        if (preferences.Channels.TryGetValue(sqsEvent.EventType, out var channels))
        {
            enabledChannels = channels;
        }
        else
        {
            var defaults = new NotificationPreference { UserId = targetUserId };
            enabledChannels = defaults.Channels.TryGetValue(sqsEvent.EventType, out var defaultChannels)
                ? defaultChannels
                : new List<DeliveryChannel> { DeliveryChannel.IN_APP };
        }

        var rendered = NotificationTemplates.Render(sqsEvent);
        var deliveredVia = new List<string>();

        if (enabledChannels.Contains(DeliveryChannel.IN_APP))
        {
            deliveredVia.Add("in_app");
        }

        if (enabledChannels.Contains(DeliveryChannel.EMAIL))
        {
            deliveredVia.Add("email");
        }

        var notification = new Notification
        {
            Id = Guid.NewGuid().ToString(),
            UserId = targetUserId,
            Type = sqsEvent.EventType,
            Title = rendered.Title,
            Message = rendered.Message,
            ResourceId = ResolveResourceId(sqsEvent),
            ResourceType = ResolveResourceType(sqsEvent),
            ActorId = string.IsNullOrEmpty(sqsEvent.ActorId) ? sqsEvent.OwnerId : sqsEvent.ActorId,
            Read = false,
            DeliveredVia = deliveredVia,
            CreatedAt = DateTime.UtcNow.ToString("O"),
        };

        await _repository.SaveNotificationAsync(notification);
        _logger.LogInformation("Stored notification {Id} for user {UserId}", notification.Id, targetUserId);

        if (enabledChannels.Contains(DeliveryChannel.PUSH))
        {
            deliveredVia.Add("push");
            await _repository.SaveNotificationAsync(new Notification
            {
                Id = notification.Id,
                UserId = notification.UserId,
                Type = notification.Type,
                Title = notification.Title,
                Message = notification.Message,
                ResourceId = notification.ResourceId,
                ResourceType = notification.ResourceType,
                ActorId = notification.ActorId,
                Read = notification.Read,
                DeliveredVia = deliveredVia,
                CreatedAt = notification.CreatedAt,
            });
        }

        _logger.LogInformation(
            "Processed {EventType} for user {UserId} via {Channels}",
            sqsEvent.EventType, targetUserId, string.Join(", ", deliveredVia));
    }

    public Task<(List<Notification> Notifications, int Total)> GetNotificationsAsync(string userId, int page, int pageSize) =>
        _repository.GetNotificationsByUserIdAsync(userId, page, pageSize);

    public Task<int> GetUnreadCountAsync(string userId) =>
        _repository.GetUnreadCountAsync(userId);

    public Task<bool> MarkAsReadAsync(string notificationId) =>
        _repository.MarkAsReadAsync(notificationId);

    public Task<int> MarkAllAsReadAsync(string userId) =>
        _repository.MarkAllAsReadAsync(userId);

    public Task<bool> DeleteNotificationAsync(string notificationId) =>
        _repository.DeleteNotificationAsync(notificationId);

    public Task<Notification?> GetNotificationByIdAsync(string id) =>
        _repository.GetNotificationByIdAsync(id);

    public Task<NotificationPreference> GetPreferencesAsync(string userId) =>
        _repository.GetPreferencesAsync(userId);

    public async Task UpdatePreferencesAsync(string userId, string eventType, List<DeliveryChannel> channels)
    {
        var current = await _repository.GetPreferencesAsync(userId);
        var updatedChannels = new Dictionary<string, List<DeliveryChannel>>(current.Channels)
        {
            [eventType] = channels,
        };

        await _repository.SavePreferencesAsync(new NotificationPreference
        {
            UserId = userId,
            Channels = updatedChannels,
        });
    }

    public static string ResolveTargetUserId(SqsNotificationMessage sqsEvent)
    {
        return sqsEvent.EventType switch
        {
            "file_shared" => sqsEvent.SharedWithUserId,
            "comment_added" => string.IsNullOrEmpty(sqsEvent.UserId) ? sqsEvent.OwnerId : sqsEvent.UserId,
            "document_edited" => string.IsNullOrEmpty(sqsEvent.UserId) ? sqsEvent.OwnerId : sqsEvent.UserId,
            "user_mentioned" => string.IsNullOrEmpty(sqsEvent.MentionedUserId) ? sqsEvent.UserId : sqsEvent.MentionedUserId,
            _ => sqsEvent.UserId,
        };
    }

    public static string ResolveResourceId(SqsNotificationMessage sqsEvent)
    {
        return sqsEvent.EventType switch
        {
            "file_shared" => sqsEvent.FileId,
            "comment_added" => string.IsNullOrEmpty(sqsEvent.CommentId) ? sqsEvent.DocumentId : sqsEvent.CommentId,
            "document_edited" => sqsEvent.DocumentId,
            "user_mentioned" => sqsEvent.DocumentId,
            _ => string.Empty,
        };
    }

    public static string ResolveResourceType(SqsNotificationMessage sqsEvent)
    {
        return sqsEvent.EventType switch
        {
            "file_shared" => "file",
            "comment_added" => "comment",
            "document_edited" => "document",
            "user_mentioned" => "document",
            _ => "unknown",
        };
    }
}

File: services/notification-service/src/main/kotlin/com/otterworks/notification/service/NotificationService.kt
package com.otterworks.notification.service

import com.otterworks.notification.model.DeliveryChannel
import com.otterworks.notification.model.EventType
import com.otterworks.notification.model.Notification
import com.otterworks.notification.model.NotificationPreference
import com.otterworks.notification.model.SqsNotificationMessage
import com.otterworks.notification.repository.NotificationRepository
import com.otterworks.notification.template.NotificationTemplates
import com.otterworks.notification.websocket.WebSocketManager
import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.MeterRegistry
import mu.KotlinLogging
import java.time.Instant
import java.util.UUID

private val logger = KotlinLogging.logger {}

class NotificationService(
    private val repository: NotificationRepository,
    private val emailSender: EmailSender,
    private val webSocketManager: WebSocketManager,
    private val meterRegistry: MeterRegistry?,
) {
    private val processedCounter: Counter? = meterRegistry?.counter("notifications.processed")
    private val emailSentCounter: Counter? = meterRegistry?.counter("notifications.email.sent")
    private val pushSentCounter: Counter? = meterRegistry?.counter("notifications.push.sent")

    suspend fun processEvent(event: SqsNotificationMessage) {
        val targetUserId = resolveTargetUserId(event)
        if (targetUserId.isBlank()) {
            logger.warn { "No target user for event: ${event.eventType}" }
            return
        }

        val preferences = repository.getPreferences(targetUserId)
        val enabledChannels = preferences.channels[event.eventType]
            ?: NotificationPreference(userId = targetUserId).channels[event.eventType]
            ?: listOf(DeliveryChannel.IN_APP)

        val rendered = NotificationTemplates.render(event)
        val deliveredVia = mutableListOf<String>()

        if (DeliveryChannel.IN_APP in enabledChannels) {
            deliveredVia.add("in_app")
        }

        // Attempt email delivery
        if (DeliveryChannel.EMAIL in enabledChannels) {
            val emailSent = emailSender.sendEmail(
                toAddress = "$targetUserId@otterworks.io",
                subject = rendered.emailSubject,
                htmlBody = rendered.emailBody,
            )
            if (emailSent) {
                deliveredVia.add("email")
                emailSentCounter?.increment()
            }
        }

        // Create notification and persist for audit trail and offline retrieval
        val notification = Notification(
            id = UUID.randomUUID().toString(),
            userId = targetUserId,
            type = event.eventType,
            title = rendered.title,
            message = rendered.message,
            resourceId = resolveResourceId(event),
            resourceType = resolveResourceType(event),
            actorId = event.actorId.ifEmpty { event.ownerId },
            read = false,
            deliveredVia = deliveredVia,
            createdAt = Instant.now().toString(),
        )

        repository.saveNotification(notification)
        logger.info { "Stored notification ${notification.id} for user $targetUserId" }

        // Attempt WebSocket push after save, then update record with actual delivery status
        if (DeliveryChannel.PUSH in enabledChannels) {
            val pushCount = webSocketManager.pushNotification(targetUserId, notification)
            if (pushCount > 0) {
                deliveredVia.add("push")
                pushSentCounter?.increment()
                repository.saveNotification(notification.copy(deliveredVia = deliveredVia))
                logger.info { "Pushed notification ${notification.id} to $pushCount session(s) for user $targetUserId" }
            }
        }

        processedCounter?.increment()
        logger.info {
            "Processed ${event.eventType} for user $targetUserId via ${deliveredVia.joinToString()}"
        }
    }

    suspend fun getNotifications(userId: String, page: Int, pageSize: Int) =
        repository.getNotificationsByUserId(userId, page, pageSize)

    suspend fun getUnreadCount(userId: String): Int =
        repository.getUnreadCount(userId)

    suspend fun markAsRead(notificationId: String): Boolean =
        repository.markAsRead(notificationId)

    suspend fun markAllAsRead(userId: String): Int =
        repository.markAllAsRead(userId)

    suspend fun deleteNotification(notificationId: String): Boolean =
        repository.deleteNotification(notificationId)

    suspend fun getNotificationById(id: String): Notification? =
        repository.getNotificationById(id)

    suspend fun getPreferences(userId: String): NotificationPreference =
        repository.getPreferences(userId)

    suspend fun updatePreferences(userId: String, eventType: String, channels: List<DeliveryChannel>) {
        val current = repository.getPreferences(userId)
        val updatedChannels = current.channels.toMutableMap()
        updatedChannels[eventType] = channels
        repository.savePreferences(current.copy(channels = updatedChannels))
    }

    companion object {
        private val FILE_SHARED = EventType.file_shared.name
        private val COMMENT_ADDED = EventType.comment_added.name
        private val DOCUMENT_EDITED = EventType.document_edited.name
        private val USER_MENTIONED = EventType.user_mentioned.name

        fun resolveTargetUserId(event: SqsNotificationMessage): String {
            return when (event.eventType) {
                FILE_SHARED -> event.sharedWithUserId
                COMMENT_ADDED -> event.userId.ifEmpty { event.ownerId }
                DOCUMENT_EDITED -> event.userId.ifEmpty { event.ownerId }
                USER_MENTIONED -> event.mentionedUserId.ifEmpty { event.userId }
                else -> event.userId
            }
        }

        fun resolveResourceId(event: SqsNotificationMessage): String {
            return when (event.eventType) {
                FILE_SHARED -> event.fileId
                COMMENT_ADDED -> event.commentId.ifEmpty { event.documentId }
                DOCUMENT_EDITED -> event.documentId
                USER_MENTIONED -> event.documentId
                else -> ""
            }
        }

        fun resolveResourceType(event: SqsNotificationMessage): String {
            return when (event.eventType) {
                FILE_SHARED -> "file"
                COMMENT_ADDED -> "comment"
                DOCUMENT_EDITED -> "document"
                USER_MENTIONED -> "document"
                else -> "unknown"
            }
        }
    }
}
